`storage`: check for exceptions in `log_manager::disk_usage()`

WillemKauf opened the pull request:
A single log should not be able to block space management of all local logs.

To avoid this existing behavior, inspect the result of `log->disk_usage()` in a `then_wrapped()` continuation within `log_manager::disk_usage()`. We will log a message and return an empty usage report if the disk usage call failed for any reason, but will allow the log manager to continue on checking the other logs on this shard.

A motivating example was a recent blockage of space management due to:

```
resource_mgmt - storage.cc:501 - Unable to collect log storage usage. Skipping management tick: std::runtime_error (ntp {kafka/topic/0}: data offset 999 is outside the translation range (starting at 1000)))
```

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.1.x
- [X] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Make space management more resilient to exceptions thrown from individual logs.

WillemKauf committed:
`storage`: check for exceptions in `log_manager::disk_usage()`

A single log should not be able to block space management of all local
logs.

To avoid this existing behavior, inspect the result of `log->disk_usage()`
in a `then_wrapped()` continuation within `log_manager::disk_usage()`.
We will log a message and return an empty usage report if the disk usage
call failed for any reason, but will allow the log manager to continue on
checking the other logs on this shard.

WillemKauf committed:
`storage`: add `disk_usage_with_log_throwing_exception`

To assert that one log which throws an exception cannot block
a call to `log_manager::disk_usage()` from completing.

andrwng commented on src/v/storage/log_manager.cc:
```diff
@@ -1019,7 +1019,20 @@ ss::future<usage_report> log_manager::disk_usage() {
       [&limit, cfg](ss::shared_ptr<log> log) {
           return ss::with_semaphore(
-            limit, 1, [cfg, log] { return log->disk_usage(cfg); });
+                   limit, 1, [cfg, log] { return log->disk_usage(cfg); })
+            .then_wrapped([log](ss::future<usage_report> f) {
+                if (f.failed()) {
+                    auto e = f.get_exception();
+                    vlog(
+                      gclog.warn,
+                      "Unable to collect disk usage from ntp {}: {}",
+                      log->config().ntp(),
+                      e);
+                    return ss::make_ready_future<usage_report>();
+                } else {
+                    return f;
+                }
+            });
```
IIUC this will mean that space management will not see any of the space as available for reclaim, but also just won't see the space at all -- is that right? So if if this failed partition were the largest offender of disk space, I think that means that space management may think there's actually no problem.

Instead, I wonder if we could update the log's `disk_usage()` call to return a report that indicates its used size, but also indicates that none of it can be reclaimed

oleiman replied:
is it given that usage (vs reclaimable) is available at this point? it's not clear to me from the code that either computation necessarily succeeds.

related q: in general, would we expect an error caught this way to be transient or permanent?

WillemKauf replied:
> Instead, I wonder if we could update the log's disk_usage() call to return a report that indicates its used size, but also indicates that none of it can be reclaimed

Added new logic in `disk_log_impl::disk_usage()` to handle this case.

WillemKauf replied:
> is it given that usage (vs reclaimable) is available at this point? it's not clear to me from the code that either computation necessarily succeeds.

calculating the usage is simply a matter of iterating over `_segs` and accumulating their size via `seg->persistent_size()`- this function _seems_ to be exception free, so this should be safe to do.

The reclaimable size is a different matter, and it is the source of the thrown exceptions in the motivating case in the cover letter.

> in general, would we expect an error caught this way to be transient or permanent?

In most cases, I would hope transient, but in the motivating case the issues with the problem logs were permanent (due to other outstanding bugs).

oleiman replied:
yeah, `segment::persisted_size` does look exception-safe to me 👍 

thought: might be nice to extend that to the rest of `disk_usage_and_reclaimable_space` and avoid duplicating the size accumulation, but maybe that's out of scope.

WillemKauf committed:
`storage`: handle exception in `disk_log_impl::disk_usage`

In the case that `disk_usage_and_reclaimable_bytes()` throws an
exception for any reason, return the usage as the persisted size on disk
with `0` bytes indicated as reclaimable.

This will help space management take into account the space used by a
partition that is throwing an exception when calculating its disk
usage statistics (even if none of it is marked as reclaimable).

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64650-->
<!--982db106-739a-458d-8317-b4b48bd709cb-->
<!--6000c400-3239-4b57-82d6-8e80140513d5-->
<!--f68c2444-aef5-4008-bd03-754d8fc411ae-->
<!--title-->
## Retry command for [Build#64650](https://buildkite.com/redpanda/redpanda/builds/64650)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/audit_log_test.py::AuditLogTestsAppLifecycle.test_recovery_mode
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_unavailable_nodes
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--22d7da52-8939-48a1-92bf-a08fdc3523a9-->
<!--TEST RESULTS-->
<!--7a593f77-c760-45b5-a2ba-d5ade7f79b9c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64650">build#64650</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.audit_log_test.AuditLogTestsAppLifecycle.test_recovery_mode | ducktape | https://buildkite.com/redpanda/redpanda/builds/64650#019641e9-edd6-4408-8ec8-4089d482169f | FLAKY | 20/21 |
| rptest.tests.controller_log_limiting_test.ControllerLogLimitMirrorMakerTests.test_mirror_maker_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/64650#01964204-9a32-465b-9bbb-5d2b281d1d06 | FLAKY | 20/21 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/64650#01964204-9a32-465b-9bbb-5d2b281d1d06 | FLAKY | 11/14 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/64650#01964204-9a30-40b2-a129-3a351e962c60 | FLAKY | 17/21 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/64650#01964204-9a31-49d0-be95-e9887096b11d | FLAKY | 17/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64650#01964204-9a31-49d0-be95-e9887096b11d | FLAKY | 13/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64650#01964204-9a32-465b-9bbb-5d2b281d1d06 | FLAKY | 4/11 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64696">build#64696</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64696#0196451a-b0b7-4e82-9422-e10f2ac86be5 | FLAKY         | 18/21    |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64696#0196451a-b0b8-482d-8810-1ad8a6d637cd | FLAKY         | 16/21    |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64696#0196451a-b0b4-486b-a382-41266a7e8874 | FLAKY         | 15/21    |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64696#0196451f-1a76-4fb9-b530-24993a69f1b8 | FLAKY         | 17/21    |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsets.test_consumer_group_offsets                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64696#0196451a-b0b6-47ea-bfcd-e7913eed8963 | FLAKY         | 17/21    |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=2                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/64696#0196451a-b0b8-482d-8810-1ad8a6d637cd | FLAKY         | 20/21    |
</details>
<!--/body-->

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -3841,13 +3841,49 @@ ss::future<usage_report> disk_log_impl::disk_usage(gc_config cfg) {
+          usage{},
+          [](usage acc, usage u) { return acc + u; });
+
+        vlog(
+          gclog.debug,
```
nit: maybe worth being a warning?

WillemKauf replied:
Fixed

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64696-->
<!--015f5ae3-c4c5-49c1-8b47-c41bf035e35e-->
<!--4a5ff138-7d9f-40a0-b64f-304adcfda8d4-->
<!--615b3cb3-101f-4f27-96ae-6e48a23a7f2e-->
<!--0f2d877e-d7c4-49ba-b28b-2573e4322e31-->
<!--b60a5080-3ff6-4ac6-8771-cd8c92f7e07c-->
<!--title-->
## Retry command for [Build#64696](https://buildkite.com/redpanda/redpanda/builds/64696)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsets.test_consumer_group_offsets
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":true}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":true},"transfer_leadership":false}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets@{"versions_to_upgrade":2}
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"nessie","cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25833-v24.3.x-995 remotes/upstream/v24.3.x
git cherry-pick -x a0dba5910e 5e7b6fa6f9 ce40ee858a
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14576447973).
